crates/core/src/masp.rs

```diff
@@ -23,6 +24,72 @@ use crate::string_encoding::{
 };
 use crate::token::{Denomination, MaspDigitPos};
 
+/// Wrapper type around `Epoch` for type safe operations involving the masp
+/// epoch
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    BorshDeserializer,
+    BorshSchema,
+    Clone,
+    Copy,
+    Debug,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+)]
+pub struct MaspEpoch(Epoch);
+
+impl Display for MaspEpoch {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for MaspEpoch {
+    type Err = ParseIntError;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let raw: u64 = u64::from_str(s)?;
+        Ok(Self(Epoch(raw)))
+    }
+}
+
+impl MaspEpoch {
+    /// Converts and `Epoch` into a `MaspEpoch` based on the provided conversion
+    /// rate
+    pub fn try_from_epoch(
+        epoch: Epoch,
+        masp_epoch_multiplier: u64,
```

Comment:
```suggestion
        masp_epoch_multiplier: NonZeroU64,
```
avoids this being a faulty op I think (perhaps not bc of div underflows)